pytorch_to_returnn/torch/nn/utils/rnn.py

```diff
@@ -37,7 +37,8 @@ def pack_sequence(sequences, enforce_sorted=True):
     batch_dim = sequences.shape[torch_axis_from_returnn_axis[tensor_entry.returnn_data.batch_dim_axis]]
     time_dim = sequences.shape[torch_axis_from_returnn_axis[tensor_entry.returnn_data.time_dim_axis]]
     lengths = [time_dim] * batch_dim
-    return pack_padded_sequence(sequences, lengths, enforce_sorted=enforce_sorted)
+    batch_first = True  # we assume batch-time-major tensor, see above
```

Comment:
I think the `batch_first` in `pack_padded_sequence` refers to its input. And here, the input is always batch major, because only in this case `pack_sequence` with a batched tensor works as expected.

Reply:
Oh really? Then we handle it incorrectly in `pack_padded_sequence`. Because we use it for the output, i.e. the packed tensor. Or maybe in PyTorch it refers to both input and output? Then it would be correct because the input should not matter at all. It should know internally what the batch dim is and handle any case correctly.

In any case, the input should not matter here at all and should not make any difference. This just has an effect on the output. And I wonder, in all usages of `PackedSequence` in PyTorch that I know (`LSTM` etc), it expects a time-major packed tensor.

Reply:
I think in `pack_padded_sequence` it refers to whether the input is a batch major tensor and in `pad_packed_sequence` it refers to whether the output should be batch major or not.

The packed tensor looks the same whether `batch_first` is `True` or `False` in  `pack_padded_sequence`. So I guess `FlattenBatch` should always use `batch_first=False`. For `pad_packed_sequence` we need to extend a bit, because PyTorch returns it according to `batch_first` (by default `False`) but our `PackedSequence.get_batched_tensor` returns it in whatever format the input before flattening was.

Reply:
Do you also think so?

Reply:
`FlattenBatch` should always use `batch_major=False` then, yes.

And I think we can just ignore the `batch_first`. Or we can make an assertion that it matches what we have in the Torch tensor.

Reply:
I think we can take care of `batch_first` in `get_batched_tensor`, see the most recent commit. Note that the initial change in this PR is removed then.

Reply:
I added this again, as `batch_first` refers to the input which is a batch major tensor here, so `batch_first` should be `True`. (It's not used but we added the assertion in `pack_padded_sequence_with_batch_sizes` now).

Reply:
I would not introduce it as a local variable put just pass `batch_first=batch_first` directly below. But with the same comment before that we assume a batch-major Torch tensor here.